File: scripts/lib/counties.ts
import type { Jurisdiction } from '@dmv/shared';

export interface County {
  fips: string;
  name: string;
  shortName: string;
  jurisdiction: Jurisdiction;
  isIndependentCity?: boolean;
  /** State FIPS prefix (used for Census API queries) */
  stateFips: '11' | '24' | '51';
  /** County portion of FIPS (last 3 digits) */
  countyFips: string;
}

export const DMV_COUNTIES: readonly County[] = [
  // District of Columbia (state FIPS 11)
  {
    fips: '11001',
    name: 'District of Columbia',
    shortName: 'DC',
    jurisdiction: 'DC',
    stateFips: '11',
    countyFips: '001',
  },

  // Maryland (state FIPS 24)
  {
    fips: '24003',
    name: 'Anne Arundel County',
    shortName: 'Anne Arundel',
    jurisdiction: 'MD',
    stateFips: '24',
    countyFips: '003',
  },
  {
    fips: '24005',
    name: 'Baltimore County',
    shortName: 'Baltimore Co.',
    jurisdiction: 'MD',
    stateFips: '24',
    countyFips: '005',
  },
  {
    fips: '24009',
    name: 'Calvert County',
    shortName: 'Calvert',
    jurisdiction: 'MD',
    stateFips: '24',
    countyFips: '009',
  },
  {
    fips: '24017',
    name: 'Charles County',
    shortName: 'Charles',
    jurisdiction: 'MD',
    stateFips: '24',
    countyFips: '017',
  },
  {
    fips: '24021',
    name: 'Frederick County',
    shortName: 'Frederick',
    jurisdiction: 'MD',
    stateFips: '24',
    countyFips: '021',
  },
  {
    fips: '24027',
    name: 'Howard County',
    shortName: 'Howard',
    jurisdiction: 'MD',
    stateFips: '24',
    countyFips: '027',
  },
  {
    fips: '24031',
    name: 'Montgomery County',
    shortName: 'Montgomery',
    jurisdiction: 'MD',
    stateFips: '24',
    countyFips: '031',
  },
  {
    fips: '24033',
    name: "Prince George's County",
    shortName: "Prince George's",
    jurisdiction: 'MD',
    stateFips: '24',
    countyFips: '033',
  },
  {
    fips: '24510',
    name: 'Baltimore city',
    shortName: 'Baltimore City',
    jurisdiction: 'MD',
    isIndependentCity: true,
    stateFips: '24',
    countyFips: '510',
  },

  // Virginia (state FIPS 51)
  {
    fips: '51013',
    name: 'Arlington County',
    shortName: 'Arlington',
    jurisdiction: 'VA',
    stateFips: '51',
    countyFips: '013',
  },
  {
    fips: '51059',
    name: 'Fairfax County',
    shortName: 'Fairfax',
    jurisdiction: 'VA',
    stateFips: '51',
    countyFips: '059',
  },
  {
    fips: '51107',
    name: 'Loudoun County',
    shortName: 'Loudoun',
    jurisdiction: 'VA',
    stateFips: '51',
    countyFips: '107',
  },
  {
    fips: '51153',
    name: 'Prince William County',
    shortName: 'Prince William',
    jurisdiction: 'VA',
    stateFips: '51',
    countyFips: '153',
  },
  {
    fips: '51177',
    name: 'Spotsylvania County',
    shortName: 'Spotsylvania',
    jurisdiction: 'VA',
    stateFips: '51',
    countyFips: '177',
  },
  {
    fips: '51179',
    name: 'Stafford County',
    shortName: 'Stafford',
    jurisdiction: 'VA',
    stateFips: '51',
    countyFips: '179',
  },
  {
    fips: '51510',
    name: 'Alexandria city',
    shortName: 'Alexandria',
    jurisdiction: 'VA',
    isIndependentCity: true,
    stateFips: '51',
    countyFips: '510',
  },
  {
    fips: '51600',
    name: 'Fairfax city',
    shortName: 'Fairfax City',
    jurisdiction: 'VA',
    isIndependentCity: true,
    stateFips: '51',
    countyFips: '600',
  },
  {
    fips: '51610',
    name: 'Falls Church city',
    shortName: 'Falls Church',
    jurisdiction: 'VA',
    isIndependentCity: true,
    stateFips: '51',
    countyFips: '610',
  },
  {
    fips: '51683',
    name: 'Manassas city',
    shortName: 'Manassas',
    jurisdiction: 'VA',
    isIndependentCity: true,
    stateFips: '51',
    countyFips: '683',
  },
  {
    fips: '51685',
    name: 'Manassas Park city',
    shortName: 'Manassas Park',
    jurisdiction: 'VA',
    isIndependentCity: true,
    stateFips: '51',
    countyFips: '685',
  },
] as const;

export const FIPS_BY_ID: ReadonlyMap<string, County> = new Map(
  DMV_COUNTIES.map((c) => [c.fips, c]),
);

export function getCounty(fips: string): County {
  const c = FIPS_BY_ID.get(fips);
  if (!c) {
    throw new Error(`Unknown FIPS: ${fips}`);
  }
  return c;
}

export function countiesByJurisdiction(j: Jurisdiction): readonly County[] {
  return DMV_COUNTIES.filter((c) => c.jurisdiction === j);
}

File: go/internal/counties/counties.go
// Package counties is the canonical DMV county list, mirroring scripts/lib/counties.ts.
package counties

import "github.com/meowmix1337/dmv_housing_hub/go/internal/types"

type County struct {
	FIPS              string             `json:"fips"`
	Name              string             `json:"name"`
	ShortName         string             `json:"shortName"`
	Jurisdiction      types.Jurisdiction `json:"jurisdiction"`
	IsIndependentCity bool               `json:"isIndependentCity,omitempty"`
	StateFips         string             `json:"stateFips"`
	CountyFips        string             `json:"countyFips"`
}

var all = []County{
	// DC
	{FIPS: "11001", Name: "District of Columbia", ShortName: "DC", Jurisdiction: types.JurisdictionDC, StateFips: "11", CountyFips: "001"},

	// MD
	{FIPS: "24003", Name: "Anne Arundel County", ShortName: "Anne Arundel", Jurisdiction: types.JurisdictionMD, StateFips: "24", CountyFips: "003"},
	{FIPS: "24005", Name: "Baltimore County", ShortName: "Baltimore Co.", Jurisdiction: types.JurisdictionMD, StateFips: "24", CountyFips: "005"},
	{FIPS: "24009", Name: "Calvert County", ShortName: "Calvert", Jurisdiction: types.JurisdictionMD, StateFips: "24", CountyFips: "009"},
	{FIPS: "24017", Name: "Charles County", ShortName: "Charles", Jurisdiction: types.JurisdictionMD, StateFips: "24", CountyFips: "017"},
	{FIPS: "24021", Name: "Frederick County", ShortName: "Frederick", Jurisdiction: types.JurisdictionMD, StateFips: "24", CountyFips: "021"},
	{FIPS: "24027", Name: "Howard County", ShortName: "Howard", Jurisdiction: types.JurisdictionMD, StateFips: "24", CountyFips: "027"},
	{FIPS: "24031", Name: "Montgomery County", ShortName: "Montgomery", Jurisdiction: types.JurisdictionMD, StateFips: "24", CountyFips: "031"},
	{FIPS: "24033", Name: "Prince George's County", ShortName: "Prince George's", Jurisdiction: types.JurisdictionMD, StateFips: "24", CountyFips: "033"},
	{FIPS: "24510", Name: "Baltimore city", ShortName: "Baltimore City", Jurisdiction: types.JurisdictionMD, IsIndependentCity: true, StateFips: "24", CountyFips: "510"},

	// VA
	{FIPS: "51013", Name: "Arlington County", ShortName: "Arlington", Jurisdiction: types.JurisdictionVA, StateFips: "51", CountyFips: "013"},
	{FIPS: "51059", Name: "Fairfax County", ShortName: "Fairfax", Jurisdiction: types.JurisdictionVA, StateFips: "51", CountyFips: "059"},
	{FIPS: "51107", Name: "Loudoun County", ShortName: "Loudoun", Jurisdiction: types.JurisdictionVA, StateFips: "51", CountyFips: "107"},
	{FIPS: "51153", Name: "Prince William County", ShortName: "Prince William", Jurisdiction: types.JurisdictionVA, StateFips: "51", CountyFips: "153"},
	{FIPS: "51177", Name: "Spotsylvania County", ShortName: "Spotsylvania", Jurisdiction: types.JurisdictionVA, StateFips: "51", CountyFips: "177"},
	{FIPS: "51179", Name: "Stafford County", ShortName: "Stafford", Jurisdiction: types.JurisdictionVA, StateFips: "51", CountyFips: "179"},
	{FIPS: "51510", Name: "Alexandria city", ShortName: "Alexandria", Jurisdiction: types.JurisdictionVA, IsIndependentCity: true, StateFips: "51", CountyFips: "510"},
	{FIPS: "51600", Name: "Fairfax city", ShortName: "Fairfax City", Jurisdiction: types.JurisdictionVA, IsIndependentCity: true, StateFips: "51", CountyFips: "600"},
	{FIPS: "51610", Name: "Falls Church city", ShortName: "Falls Church", Jurisdiction: types.JurisdictionVA, IsIndependentCity: true, StateFips: "51", CountyFips: "610"},
	{FIPS: "51683", Name: "Manassas city", ShortName: "Manassas", Jurisdiction: types.JurisdictionVA, IsIndependentCity: true, StateFips: "51", CountyFips: "683"},
	{FIPS: "51685", Name: "Manassas Park city", ShortName: "Manassas Park", Jurisdiction: types.JurisdictionVA, IsIndependentCity: true, StateFips: "51", CountyFips: "685"},
}

func All() []County {
	out := make([]County, len(all))
	copy(out, all)
	return out
}

func ByFIPS(fips string) (County, bool) {
	for _, c := range all {
		if c.FIPS == fips {
			return c, true
		}
	}
	return County{}, false
}

func ByName(name string) (County, bool) {
	for _, c := range all {
		if c.Name == name {
			return c, true
		}
	}
	return County{}, false
}

func ByJurisdiction(j types.Jurisdiction) []County {
	out := make([]County, 0)
	for _, c := range all {
		if c.Jurisdiction == j {
			out = append(out, c)
		}
	}
	return out
}
